ncwriter/schema.py

```diff
@@ -18,28 +17,53 @@
 # Create a new validator class (based on Draft4Validator) to allow templates to use
 # * Python types or numpy dtypes to specify variable data types; and
 # * numpy arrays to specify variable data.
-meta = deepcopy(Draft4Validator.META_SCHEMA)
-meta['definitions']['simpleTypes']['enum'].append('datatype')
-types = deepcopy(Draft4Validator.DEFAULT_TYPES)
-types['datatype'] = (basestring, type, np.dtype)
-types['array'] = (list, np.ndarray)
-TemplateValidator = validators.create(meta_schema=meta,
-                                      validators=Draft4Validator.VALIDATORS,
-                                      default_types=types
-                                      )
+
+TemplateValidator = validators.create(meta_schema=Draft4Validator.META_SCHEMA,
+                                      validators=Draft4Validator.VALIDATORS)
+format_checker = FormatChecker()
+
+TYPE_STRINGS = ('int', 'int8', 'int16', 'int32', 'int64',
+                'float', 'float8', 'float16', 'float32', 'float64', 'double',
+                'S1', 'c', 'i1', 'b', 'B', 'u1', 'i2', 'h', 's', 'u2', 'i4', 'i', 'l', 'u4',
+                'i8', 'u8', 'f4', 'f', 'f8', 'd'
+                )
+
+
+@format_checker.checks('datatype')
+def is_datatype(value):
+    """Return whether the given value is a valid data type specification for a NetCDF variable"""
+    if isinstance(value, basestring):
+        if value in TYPE_STRINGS:
+            return True
+        else:
+            return False
+    if isinstance(value, np.dtype):
+        return True
+    if isinstance(value, type):
+        if issubclass(value, np.number):
+            return True
+        else:
+            return False
+
+    return False
+
+
+TYPES = {'array': (list, np.ndarray)}
 
 TEMPLATE_SCHEMA_JSON = resource_filename(__name__, 'template_schema.json')
 with open(TEMPLATE_SCHEMA_JSON) as f:
     TEMPLATE_SCHEMA = json.load(f)
 TemplateValidator.check_schema(TEMPLATE_SCHEMA)
 
+template_validator = TemplateValidator(TEMPLATE_SCHEMA, types=TYPES, format_checker=format_checker)
 
 GLOBAL_ATTRIBUTES_SCHEMA = TEMPLATE_SCHEMA.copy()
 GLOBAL_ATTRIBUTES_SCHEMA.pop('properties')  # remove special properties, leaving only global attributes
```

Comment:
I feel like there has to be a native way to handle this without a copy/pop but I don't know what that is! This is the only instance of fiddling with the JSON from Python, so it may become clear in future how to manage this...

Reply:
Yeah, I know it's a bit hacky. I do have a couple of ideas for avoiding this. I'll try them out.